File: 1838. Frequency of the Most Frequent Element.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        nums.sort(reverse=True)
        start, end = 0, 0
        maxfreq = 1
        
        while end < len(nums):
            if nums[start] - nums[end] > k:
                maxfreq = max(maxfreq, end - start)
                k += (end - start - 1) * (nums[start] - nums[start + 1])
                start += 1
            else:
                k -= nums[start] - nums[end]
                end += 1
            maxfreq = max(maxfreq, end - start)
        return maxfreq

File: 1838-frequency-of-the-most-frequent-element/1838-frequency-of-the-most-frequent-element.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        // when update start 
        // k += (end - start) * abs(nums[start] - nums[start + 1])
        
        int start = 0, ans = 0, n = nums.size();
        sort(nums.begin(), nums.end(), greater<>());
        
        for (int end = 0; end < n; end++) {
            while (nums[start] - nums[end] > k) {
                k += (end - start - 1) * (nums[start] - nums[start + 1]);
                start++;
            }
            ans = max(ans, end - start + 1);
            k -= nums[start] - nums[end];
        }
        return ans;
    }
};